File: tools/archetype/SpriteAtlas.py
from . import Sprite


def params_constructor(sprite_atlas, name) -> str:
    c = Sprite.params_constructor(sprite_atlas['sprite'], f'{name}.sprite_params')

    if 'starting_frame' in sprite_atlas:
        c += f"\t\t\t{name}.starting_frame = (GLuint){sprite_atlas['starting_frame']}"
    if 'starting_time' in sprite_atlas:
        c += f"\t\t\t{name}.starting_time = (float){sprite_atlas['starting_time']}"

    if 'rows' in sprite_atlas and 'cols' in sprite_atlas and 'delay_seconds' in sprite_atlas:
        c += "\t\t\t{\n"
        c += f"\t\t\t\treg::params::SpriteAtlas::Frame frame{{ .rows = (GLuint){sprite_atlas['rows']}, .cols = (GLuint){sprite_atlas['cols']}, .delay_seconds = (float){sprite_atlas['delay_seconds']} }};\n"
        if 'row_major' in sprite_atlas:
            c += f"\t\t\t\tframe.row_major = {'true' if sprite_atlas['row_major'] else 'false'};\n"
        if 'row_up' in sprite_atlas:
            c += f"\t\t\t\tframe.row_up = {'true' if sprite_atlas['row_up'] else 'false'};\n"
        c += f"\t\t\t\t{name}.frame = frame;\n"
        c += "\t\t\t}\n"
    elif 'static_frame' in sprite_atlas:
        c += f"\t\t\t{name}.frame = reg::params::SpriteAtlas::StaticFrame{{ (GLuint){sprite_atlas['static_frame']} }};\n"

    return c
    

def constructor(sprite_atlas) -> str:
    return f"""
        {{
            reg::params::SpriteAtlas params;
{params_constructor(sprite_atlas, 'params')}
            {sprite_atlas['name']}.init(reg::load_sprite_atlas(params));
        }}"""

File: engine/graphics/extensions/SpriteAtlas.cpp
#include "SpriteAtlas.h"

#include "core/util/Time.h"

namespace oly::rendering
{
	SpriteAtlas::SpriteAtlas(Sprite&& sprite)
		: sprite(std::move(sprite))
	{
	}

	void SpriteAtlas::draw() const
	{
		sprite.draw();
	}

	void SpriteAtlas::on_tick() const
	{
		if (anim_format.delay_seconds != 0.0f)
			select(anim_format.starting_frame + (int)floor((TIME.now<float>() - anim_format.starting_time) / anim_format.delay_seconds));
	}

	void SpriteAtlas::select_static_frame(GLuint frame)
	{
		anim_format.delay_seconds = 0.0f;
		select(frame);
	}

	void SpriteAtlas::uvs_changed() const
	{
		current_frame = -1;
	}

	void SpriteAtlas::setup_uniform(GLuint rows, GLuint cols, float delay_seconds, bool row_major, bool row_up)
	{
		atlas.clear();
		static const auto uv_rect = [](GLuint row, GLuint col, GLuint rows, GLuint cols) -> UVRect {
			return { {
				{       col / (float)cols,       row / (float)rows },
				{ (col + 1) / (float)cols,       row / (float)rows },
				{ (col + 1) / (float)cols, (row + 1) / (float)rows },
				{       col / (float)cols, (row + 1) / (float)rows }
			} };
			};
		if (row_major)
		{
			if (row_up)
			{
				for (GLuint row = 0; row < rows; ++row)
					for (GLuint col = 0; col < cols; ++col)
						atlas.push_back(uv_rect(row, col, rows, cols));
			}
			else
			{
				for (int row = rows - 1; row >= 0; --row)
					for (GLuint col = 0; col < cols; ++col)
						atlas.push_back(uv_rect(row, col, rows, cols));
			}
		}
		else
		{
			if (row_up)
			{
				for (GLuint col = 0; col < cols; ++col)
					for (GLuint row = 0; row < rows; ++row)
						atlas.push_back(uv_rect(row, col, rows, cols));
			}
			else
			{
				for (GLuint col = 0; col < cols; ++col)
					for (int row = rows - 1; row >= 0; --row)
						atlas.push_back(uv_rect(row, col, rows, cols));
			}
		}
		uvs_changed();
		anim_format.num_frames = rows * cols;
		anim_format.delay_seconds = delay_seconds;
		anim_format.starting_frame = 0;
		anim_format.starting_time = 0.0f;

		glm::vec2 dimensions;
		auto texture = sprite.get_texture(dimensions);
		sprite.set_texture(texture, dimensions * glm::vec2{ 1.0f / cols, 1.0f / rows });
	}

	void SpriteAtlas::select(GLuint frame) const
	{
		frame %= anim_format.num_frames;
		if (current_frame != frame)
		{
			current_frame = frame;
			sprite.set_tex_coords(atlas[frame]);
		}
	}
}
